File: src/syntax/scanner.go
package syntax

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"

	"chai/logging"
)

// NewScanner creates a scanner for the given file
func NewScanner(fpath string, lctx *logging.LogContext) (*Scanner, bool) {
	f, err := os.Open(fpath)

	if err != nil {
		logging.LogConfigError("File", "error opening file: "+err.Error())
		return nil, false
	}

	s := &Scanner{fh: f, file: bufio.NewReader(f), fpath: fpath, line: 1, lctx: lctx}
	return s, true
}

// IsLetter tests if a rune is an ASCII character
func IsLetter(r rune) bool {
	return r > '`' && r < '{' || r > '@' && r < '[' // avoid using <= and >= by checking characters on boundaries (same for IsDigit)
}

// IsDigit tests if a rune is an ASCII digit
func IsDigit(r rune) bool {
	return r > '/' && r < ':'
}

// Scanner works like an io.Reader for a file (outputting tokens)
type Scanner struct {
	lctx *logging.LogContext

	fh    *os.File
	file  *bufio.Reader
	fpath string

	line int
	col  int

	tokBuilder strings.Builder

	curr rune

	// lookahead is a token that was processed while scanning in another token
	// This token is set during one call to ReadToken() and outputted the next.
	lookahead *Token
}

// ReadToken reads a single token from the stream.  True indicates that there is
// a token to be read/processed
func (s *Scanner) ReadToken() (*Token, bool) {
	// check the lookahead before returning
	if s.lookahead != nil {
		tok := s.lookahead
		s.lookahead = nil
		return tok, true
	}

	for s.readNext() {
		tok := &Token{}
		malformed := false

		switch s.curr {
		//  ignore non-meaningful characters (eg. BOM, form-feeds, spaces etc.)
		case '\r', '\f', '\v', ' ', '\t', 65279:
			s.tokBuilder.Reset()
			continue
		// handle newlines where they are relevant
		case '\n':
			// line counting handled in readNext
			tok = s.getToken(NEWLINE)
			s.tokBuilder.Reset()

			return tok, true
		// handle string-like
		case '"':
			// trim off leading `"`
			s.tokBuilder.Reset()
			tok, malformed = s.readStdStringLiteral()
		case '\'':
			// trim off leading `'`
			s.tokBuilder.Reset()
			tok, malformed = s.readRuneLiteral()
		case '`':
			// trim off leading ```
			s.tokBuilder.Reset()
			tok, malformed = s.readRawStringLiteral()
		// handle comments
		case '#':
			ahead, more := s.peek()

			if !more {
				// some weird error happened -- not just an eof
				return nil, false
			} else if ahead == '!' {
				s.tokBuilder.Reset() // get rid of lingering `#`
				s.skipBlockComment()
				continue
			} else {
				s.tokBuilder.Reset() // get rid of lingering `#`

				// skip the line comment (by skipping until a newline)
				for s.skipNext() && s.curr != '\n' {
				}

				// return a newline so that the parser can still count newlines
				// at the end of line comments as ending lines
				return &Token{Kind: NEWLINE, Value: "\n", Line: s.line, Col: s.col}, true
			}
		// handle the split-join character ('\')
		case '\\':
			// read through any whitespace until a newline is encountered. We
			// can use regular readNext since we want to include the newline.
		loop:
			for s.readNext() {
				switch s.curr {
				// BOM can also be ignored if it occurs here
				case '\r', '\f', '\v', 65279, ' ', '\t':
					continue
				case '\n':
					break loop
				default:
					// if we encounter something that is not a whitespace
					// character before we encounter a newline, the character is
					// invalid and we mark only that character as erroneous
					logging.LogCompileError(
						s.lctx,
						"expecting a line break before next non-whitespace character",
						logging.LMKToken,
						&logging.TextPosition{StartLn: s.line, StartCol: s.col - 1, EndLn: s.line, EndCol: s.col},
					)
					return nil, false
				}
			}

			// if we reach this point, the newline has been consumed and our
			// split-join is valid.  To apply its effects, we simply discard the
			// current builder contents including the newline and continue on as
			// normal.  Since from the scanner's POV, we are not on a separate
			// line (it does count it, but doesn't act on it), it will not be
			// checking for indentation tokens and so the user can indent or
			// dedent as much as they want so long as they return to correct
			// indentation level after the next, non-split-join newline.
			s.tokBuilder.Reset()
			continue
		default:
			// check for identifiers
			if IsLetter(s.curr) || s.curr == '_' {
				tok = s.readWord()
			} else if IsDigit(s.curr) {
				// check numeric literals
				tok, malformed = s.readNumberLiteral()
			} else if kind, ok := symbolPatterns[string(s.curr)]; ok {
				// all compound tokens begin with valid single tokens so the
				// check above will match the start of any symbolic token

				// keep reading as long as our lookahead is valid: avoids
				// reading extra tokens (ie. in place of readNext)
				for ahead, more := s.peek(); more; ahead, more = s.peek() {
					if skind, ok := symbolPatterns[s.tokBuilder.String()+string(ahead)]; ok {
						kind = skind
						s.readNext()
					} else {
						break
					}
				}

				// turn whatever we managed to read into a token
				tok = s.getToken(kind)
			} else {
				// any other token must be malformed in some way
				malformed = true
			}
		}

		// discard the built contents for the current scanned token
		s.tokBuilder.Reset()

		// error out on any malformed tokens (using contents of token builder)
		if malformed {
			logging.LogCompileError(
				s.lctx,
				fmt.Sprintf("malformed token: `%s`", s.tokBuilder.String()),
				logging.LMKToken,
				&logging.TextPosition{StartLn: s.line, StartCol: s.col, EndLn: s.line, EndCol: s.col + s.tokBuilder.Len()},
			)
			return nil, false
		}

		return tok, true
	}

	// if we reach here, we are at the end of the file.  First, return a newline
	// since an end of file can be counted as one and store the EOF in the
	// lookahead
	s.lookahead = &Token{Kind: EOF}
	return s.makeToken(NEWLINE, ""), true
}

// UnreadToken is used to undo the preprocessor read the occurs at the start of
// every file.  This function should ONLY be called at the start of a file
func (s *Scanner) UnreadToken(tok *Token) {
	// lookahead should always be empty when this function is called
	s.lookahead = tok
}

// Context returns the scanner's log context
func (s *Scanner) Context() *logging.LogContext {
	return s.lctx
}

// Close closes the open file handle the scanner is processing
func (s *Scanner) Close() error {
	return s.fh.Close()
}

// create a token at the current position from the provided data
func (s *Scanner) makeToken(kind int, value string) *Token {
	tok := &Token{Kind: kind, Value: value, Line: s.line, Col: s.col}
	return tok
}

// collect the contents of the token builder into a string and create a token at
// the current position with the provided kind and token string as its value
func (s *Scanner) getToken(kind int) *Token {
	tokValue := s.tokBuilder.String()
	return s.makeToken(kind, tokValue)
}

// reads a rune from the file stream into the token builder and returns whether
// or not there are more runes to be read (true = no EOF, false = EOF)
func (s *Scanner) readNext() bool {
	r, _, err := s.file.ReadRune()

	if err != nil {
		if err != io.EOF {
			logging.LogConfigError("File", fmt.Sprintf("error reading file %s: %s", s.fpath, err.Error()))
		}

		return false
	}

	// do line and column counting after the newline token
	// as been processed (so as to avoid positioning errors)
	if s.curr == '\n' {
		s.line++
		s.col = 0
	}

	s.tokBuilder.WriteRune(r)
	s.curr = r

	if r == '\t' {
		// Whirlwind makes the executive decision to count tabs as four spaces
		// (this is only really meaningful for display purposes -- Whirlwind
		// highlights the error)
		s.col += 4
	} else {
		s.col++
	}

	return true
}

// same behavior as readNext but doesn't populate the token builder used for
// comments where it makes sense
func (s *Scanner) skipNext() bool {
	r, _, err := s.file.ReadRune()

	if err != nil {
		if err != io.EOF {
			logging.LogConfigError("File", fmt.Sprintf("error reading file %s: %s", s.fpath, err.Error()))
		}

		return false
	}

	// do line and column counting after the newline token
	// as been processed (so as to avoid positioning errors)
	if s.curr == '\n' {
		s.line++
		s.col = 0
	} else {
		s.col++
	}

	s.curr = r
	return true
}

// peek a rune ahead on the scanner (used to test for malformed tokens)
func (s *Scanner) peek() (rune, bool) {
	r, _, err := s.file.ReadRune()

	if err != nil {
		return 0, false
	}

	s.file.UnreadRune()

	return r, true
}

// reads an identifier or a keyword from the input stream determines based on
// contents of stream (matches to all possible keywords)
func (s *Scanner) readWord() *Token {
	// if our word starts with an '_', it cannot be a keyword (simple check here)
	keywordValid := s.curr != '_'

	// to read a word, we assume that current character is valid and already in
	// the token builder (guaranteed by caller or previous loop cycle). we then
	// use a look-ahead to check if the next token will be valid. If it is, we
	// continue looping (and the logic outlined above holds). If not, we exit.
	// Additionally, if at any point in the middle of the word, we encounter an
	// underscore, we know we are not reading a keyword and set the
	// corresponding flag.  This function is never called on words that begin
	// with numbers so no need to check for first-character rules in it.
	for {
		c, more := s.peek()

		if !more {
			break
		} else if c == '_' {
			keywordValid = false
		} else if !IsLetter(c) && !IsDigit(c) {
			break
		}

		s.readNext()
	}

	tokValue := s.tokBuilder.String()

	// if a keyword is possible and our current token value matches a keyword
	// pattern, create a new keyword token from the token builder
	if keywordValid {
		if kind, ok := keywordPatterns[tokValue]; ok {
			return s.makeToken(kind, tokValue)
		}
	}

	// otherwise, assume that it is just an identifier and act accordingly
	return s.makeToken(IDENTIFIER, tokValue)
}

// read in a floating point or integral number
func (s *Scanner) readNumberLiteral() (*Token, bool) {
	var isHex, isBin, isOct, isFloat, isUns, isLong bool

	// the first thing to do it to determine what kind of numeric literal we are
	// dealing with.  If our token starts with at `0`, then we peek ahead to see
	// if we need to process an integer literal prefix (ie. a 0x).  If it
	// doesn't, we do nothing and leave the next character to be scanned by the
	// main integer literal scanning loop (s.curr will be our starting char)
	if s.curr == '0' {
		ahead, more := s.peek()

		if more {
			switch ahead {
			case 'x':
				isHex = true
				s.readNext()
			case 'o':
				isOct = true
				s.readNext()
			case 'b':
				isBin = true
				s.readNext()
			}
		} else {
			// if we are out of tokens, then 0 is our only element
			return s.getToken(INTLIT), true
		}
	}

	// if we previous was an 'e' then we can expect a '-'
	expectNeg := false

	// if previous was a `.` then we expect a digit
	expectDigit := false

	// if we triggered a floating point using '.' instead of 'e' than 'e' could
	// still be valid
	eValid := false

	// use loop break label to break out loop from within switch case
loop:
	// We already know the first character is valid so we can simply ignore it.
	// Moreover, we don't need to check readNext since we know the next token
	// will be valid b/c it will be checked with s.peek() is called.  So instead
	// we just want to make sure that all `continue` because the checked
	// character to be read in.
	for ; true; s.readNext() {
		ahead, more := s.peek()

		// if we have nothing left, then we are out of tokens and should exit
		if !more {
			break
		}

		// if we have identified signage or sign, then we are not expecting
		// anymore values and so exit out if an additional values are
		// encountered besides sign and size specifiers
		if isLong && isUns {
			break
		} else if isLong {
			if ahead == 'u' {
				isUns = true
				continue
			} else {
				break
			}
		} else if isUns {
			if ahead == 'l' {
				isLong = true
				continue
			} else {
				break
			}
		}

		// if we are expecting a negative and get another character then we
		// simply update the state (no longer expecting a negative) and continue
		// on (expect is not a hard expectation)
		if expectNeg && ahead != '-' {
			expectNeg = false
		}

		// check to ensure that any binary literals are valid
		if isBin {
			if ahead == '0' || ahead == '1' {
				continue
			} else {
				break
			}
		}

		// check to ensure that any octal literals are valid
		if isOct {
			if ahead > '/' && ahead < '9' {
				continue
			} else {
				break
			}
		}

		if IsDigit(ahead) {
			expectDigit = false
			continue
		}

		// check for validity of hex literal
		if isHex && (ahead < 'A' || ahead > 'F') && (ahead < 'a' || ahead > 'f') {
			break
		} else if isFloat {
			// if we are expecting a digit and not getting one (ie. after the `.`)
			if expectDigit {
				// create the base integer literal
				tokValue := s.tokBuilder.String()
				intlit := s.makeToken(INTLIT, tokValue[:len(tokValue)-1])
				intlit.Col--

				// clear the token builder before we continue to build the dots
				s.tokBuilder.Reset()

				// accumulate `.` into larger tokens if necessary
				for ahead == '.' {
					s.readNext()
					ahead, more = s.peek()

					if !more {
						break
					}
				}

				// make the dots token into the current lookahead
				s.lookahead = s.getToken(symbolPatterns[s.tokBuilder.String()])
				s.tokBuilder.Reset()

				return intlit, false
			}

			// our scanning logic changes after we parse a float: we can longer
			// accept `u`, `l` or `.` (since even if an `e` (or `E`) was what
			// moved us into this state, the power must be an integer value).
			// So we employ alternative checking logic
			switch ahead {
			case 'e', 'E':
				// avoid duplicate e's
				if eValid {
					eValid = false
				} else {
					break loop
				}
			case '-':
				if expectNeg {
					// since we are already looking ahead, we have to assume that
					// the negative is part of this token.  If what comes after this
					// token is not a digit, then we would have something of the form
					// `0e-...`` which can never be valid even if you split the tokens
					// (b/c you get `0e`, `-`, and `...` and the `0e` is invalid)
					s.readNext()

					// check if there is a non-number ahead then we actually
					// have 3 tokens and have to scan the other two separately
					ahead, valid := s.peek()

					// hit EOF on peek, malformed token
					if !valid {
						return nil, true
					}

					// if it is not a digit, the token is malformed
					if !IsDigit(ahead) {
						return nil, true
					}

					expectNeg = false
				} else {
					break loop
				}
			default:
				// if expectNeg is true here, then we know the previous
				// character we read in was an `e` and since a numeric literal
				// cannot end with an `e` and there will never be context where
				// an identifier followed by an integer literal could be two
				// valid tokens (according to the grammar), we mark such
				// literals as one malformed token
				if expectNeg {
					return nil, true
				}

				break loop
			}
		}

		// we only reach this point, we are not a floating point value (and
		// might want to become one or need to act knowing we aren't one)
		switch ahead {
		case '.':
			isFloat = true
			eValid = true
			expectDigit = true
		case 'e', 'E':
			isFloat = true
			expectNeg = true
		case 'u':
			if isFloat {
				break
			}

			isUns = true
		case 'l':
			if isFloat {
				break
			}

			isLong = true
		default:
			break loop
		}
	}

	// binary, octal, decimal, and hexadecimal literals are all considered
	// integer literals and so the only decision here is whether or not to
	// create a floating point literal (use already accumulated information)
	if isFloat {
		return s.getToken(FLOATLIT), false
	}

	return s.getToken(INTLIT), false
}

// read in a standard string literal -- assuming leading `"` has been dropped
func (s *Scanner) readStdStringLiteral() (*Token, bool) {
	expectingEscape := false

	// use a lookahead pattern to avoid reading the closing quote
	for next, ok := s.peek(); ok; next, ok = s.peek() {
		// test for escape first
		if expectingEscape {
			// handle invalid escape sequences -- no need to read next here
			// since our `readEscapeSequence` does that for us
			if s.readEscapeSequence() {
				expectingEscape = false
			} else {
				return nil, true
			}
		}

		switch next {
		case '\\':
			expectingEscape = true
			s.readNext()
		case '"':
			// we don't want to read the closing quote, so we skip it (so that
			// the end column of the string is accurate to to where it actually
			// ends), make our string token, and return.  We know that this
			// quote is valid since escape sequences are handled above this
			// switch
			s.skipNext()
			return s.getToken(STRINGLIT), false
		case '\n':
			// catch newlines in strings
			return nil, true
		default:
			// otherwise, just read in the next character
			s.readNext()
		}
	}

	// if we reach here, we didn't encounter a proper closing quotation and
	// therefore, we need to indicate that this token is malformed
	return nil, true
}

// read in a rune literal -- assuming leading `'` has been dropped
func (s *Scanner) readRuneLiteral() (*Token, bool) {
	// if the rune has no content then it is malformed
	if !s.readNext() {
		return nil, true
	}

	// if there is an escape sequence, read it and if it is invalid, rune lit is
	// malformed
	if s.curr == '\\' && !s.readEscapeSequence() {
		return nil, true
	}

	// if the next token after processing the escape sequence/rune body is not a
	// closing quote than the rune literal is too long on we are at EOF =>
	// malformed in either case
	if next, ok := s.peek(); !ok || next != '\'' {
		return nil, true
	}

	// assume it is properly formed and skip the closing single quote; skip
	// before creating the token so the column is correct
	s.skipNext()
	return s.getToken(RUNELIT), false
}

// read and interpret an escape sequence inside a string or rune literal
func (s *Scanner) readEscapeSequence() bool {
	if !s.readNext() {
		return false
	}

	readUnicodeSequence := func(count int) bool {
		for i := 0; i < count; i++ {
			if !s.readNext() {
				return false
			}

			r := s.curr

			if !IsDigit(r) && (r < 'A' || r > 'F') && (r < 'a' || r > 'f') {
				return false
			}
		}

		return true
	}

	switch s.curr {
	case 'a', 'b', 'n', 'f', 'r', 't', 'v', '0', 's', '"', '\'', '\\':
		return true
	case 'x':
		return readUnicodeSequence(2)
	case 'u':
		return readUnicodeSequence(4)
	case 'U':
		return readUnicodeSequence(8)
	}

	return true
}

// read in a raw string literal -- assuming leading backtick has been dropped
func (s *Scanner) readRawStringLiteral() (*Token, bool) {
	// used to signal when we need to escape a backtick
	escapeNext := true

	// use a lookahead pattern to avoid reading the closing backtick
	for next, ok := s.peek(); ok; next, ok = s.peek() {
		if next != '`' || escapeNext {
			s.readNext()
		} else {
			// skip closing backtick; creating stringlit token after skipping to
			// generate correct column position
			s.skipNext()
			return s.getToken(STRINGLIT), false
		}

		// check for escapes for backticks; update after we have handled the
		// exit condition
		escapeNext = next == '\\'
	}

	// only stringlits that reach here are incomplete
	return nil, true
}

func (s *Scanner) skipBlockComment() {
	// skip opening '!'
	s.skipNext()

	for s.skipNext() {
		if s.curr == '!' {
			p, more := s.peek()

			if more && p == '#' {
				s.skipNext()
				return
			}
		}
	}
}

File: src/tokenize/scanner.cpp
#include "scanner.hpp"

#include <stdexcept>
#include <format>
#include <locale>
#include <codecvt>

#include "patterns.hpp"
#include "report/message.hpp"

namespace chai {
    Scanner::Scanner(const std::string& fpath) 
        : line(1)
        , col(1)
        , fpath(fpath)
    {
        file.open(fpath);
        if (!file)
            throw std::runtime_error(std::format("unable to open file at {}", fpath));        
    }

    Token Scanner::scanNext() {
        while (auto oc = peekChar()) {
            switch (oc.value()) {
                // skip all non-meaningful whitespace characters
                case '\t':
                case ' ':
                case '\r':
                case '\v': 
                case '\f':
                    skipChar();
                    break;
                // handle newlines
                case '\n':
                    readChar();
                    return makeToken(TokenKind::Newline);
                // handle comments
                case '#':
                    skipComment();
                    break;
                // handle split join
                case '\\':
                    skipChar();

                    if (auto ahead = peekChar()) {
                        if (ahead == '\n')
                            skipChar();
                        else {
                            throwScanError("expecting newline after split join");
                        }
                    }

                    break;
                // handle rune literals
                case '\'':
                    readChar();
                    return scanRuneLit();
                // handle std string literals
                case '\"':
                    readChar();
                    return scanStdStringLit();
                // handle raw string literals
                case '`':
                    readChar();
                    return scanRawStringLit();
                default:
                    // numeric constant
                    if (isdigit(oc.value()))
                        return scanNumberLit(oc.value());
                    // identifier or keyword
                    else if (iscsymf(oc.value())) {
                        // read in as long as the next character is valid
                        do {
                            readChar();
                        } while ((oc = peekChar()) && iscsym(oc.value()));

                        // check if the identifier matches a known keyword and
                        // return an appropriate token if so
                        if (auto kind = matchKeyword(tokBuff))
                            return makeToken(kind.value());

                        // if not, return it as an identifier
                        return makeToken(TokenKind::Identifier);
                    }
                    // operator or punctuation
                    else {
                        TokenKind kind = TokenKind::EndOfFile;
                        std::optional<TokenKind> okind;
                        while ((oc = peekChar()) && (okind = matchSymbol(tokBuff + oc.value()))) {
                            readChar();
                            kind = okind.value();
                        }                           

                        // kind == EOF => first token didn't match a symbol
                        if (kind == TokenKind::EndOfFile) {
                            readChar();
                            throwScanError("unexpected character");
                        }
                        
                        // otherwise it is a valid token :)
                        return makeToken(kind);
                    }
                    break;
            }
        }

        // if we reach here, this is always an EOF
        return {.kind = TokenKind::EndOfFile, .line=line, .col=col};
    }

    void Scanner::skipComment() {
        skipChar();

        if (auto ahead = peekChar()) {
            // multi-line comment
            if (ahead == '!') {
                skipChar();

                while (auto ahead = peekChar()) {
                    // discard the character
                    skipChar();

                    // if we hit an exclamation point, check if what
                    // follows is a `#`.  if so, exit the comment
                    if (ahead == '!') {
                        if ((ahead = peekChar()) && ahead == '#') {
                            skipChar();
                            break;
                        }
                    }
                }
            }
            // single-line comment 
            else {
                while ((ahead = peekChar()) && ahead != '\n') {
                    skipChar();
                }         
            }
        }
    }

    // scanRuneLit scans in a rune literal assuming the leading single quote has
    // already been read in.
    Token Scanner::scanRuneLit() {
        if (auto ahead = peekChar()) {
            if (ahead == '\\') {
                skipChar();

                readEscapeSequence();
            } else if (ahead == '\'')
                throwScanError("empty rune literal");
            else readChar();

            if ((ahead = peekChar()) && ahead == '\'') {
                readChar();

                return makeToken(TokenKind::RuneLiteral);
            }
        }

        // missing the closing quote
        throwScanError("expected closing quote for rune literal");
    }

    // readEscapeSequence is used to read in and interpret the sequence of
    // characters following a backslash in a string or rune.  It assumes the
    // leading backslash has already been skipped.  The correct "value" of the
    // escape sequence is inserted into the token buffer.
    void Scanner::readEscapeSequence() {
        if (auto c = peekChar()) {
            skipChar();
            switch (c.value()) {
                case 'a':
                    tokBuff.push_back('\a');
                    break;
                case 'b':
                    tokBuff.push_back('\b');
                    break;
                case 'f':
                    tokBuff.push_back('\f');
                    break;
                case 'n':
                    tokBuff.push_back('\n');
                    break;
                case 'r':
                    tokBuff.push_back('\r');
                    break;
                case 't':
                    tokBuff.push_back('\t');
                    break;
                case 'v':
                    tokBuff.push_back('\v');
                    break;
                case '0':
                    tokBuff.push_back('\0');
                    break;
                case '\'':
                    tokBuff.push_back('\'');
                    break;
                case '\"':
                    tokBuff.push_back('\"');
                    break;
                case '\\':
                    tokBuff.push_back('\\');
                    break;
                // special utf-8 encodings
                case 'x':
                    readUnicodeSequence(2);
                    break;
                case 'u':
                    readUnicodeSequence(4);
                    break;
                case 'U':
                    readUnicodeSequence(8);
                    break;
                default:
                    throwScanError("unknown escape code");
                    break;
            }
        } else
            throwScanError("expected escape sequence following backslash in literal");
    }

    // readUnicodeSequence reads a sequence of hex-encoded values and interprets
    // them as a single unicode code point.  This point is then converted into
    // UTF-8 bytes and written to the token buffer.  The input determines the
    // expected length of the unicode sequence.
    void Scanner::readUnicodeSequence(int count) {
        // read in the raw unicode string
        std::string unicodeStr;
        for (int i = 0; i < count; i++) {
            if (auto ahead = peekChar()) {
                if (isdigit(ahead.value()) || 'A' <= ahead && ahead < 'G' || 'a' <= ahead && ahead < 'g') {
                    unicodeStr.push_back(ahead.value());
                    skipChar();
                }
                else
                    throwScanError(std::format("expected a hexadecimal character (0-9, a-f, or A-F) not {}", ahead.value()));
            } else
                throwScanError("unexpected end of unicode escape sequence");
        }

        // convert the unicode string to an 32-bit unicode point
        char32_t codePoint = std::stoul(unicodeStr, nullptr, 16);

        // convert the unicode point to utf-8 bytes
        std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> converter;

        try {
            std::string u8bytes = converter.to_bytes(codePoint);

            // write the bytes to the token buffer
            tokBuff += u8bytes;
        } catch (std::exception) {
            throwScanError(std::format("{} is not a valid unicode character", unicodeStr));
        }      
    }

    // scanNumberLit scans in a numeric literal.
    Token Scanner::scanNumberLit(char c) {
        // read in the leading number
        readChar();

        int base = 10;

        // first, check for alternative bases -- note, we do include the leading
        // zero even in the case of a variable base so that backend knows what
        // kind of number literal to output to LLVM
        if (c == '0') {
            if (auto ahead = peekChar()) {
                // if the next character is a digit, it is obviously not a
                // variable base so we keep reading
                if (!isdigit(ahead.value())) {
                    switch (ahead.value()) {
                        // handle the three base codes
                        case 'b':
                            readChar();
                            base = 2;
                            break;
                        case 'o':
                            readChar();
                            base = 8;
                            break;
                        case 'x':
                            readChar();
                            base = 16;
                            break;
                        // handle all non-digit, but valid components of numeric
                        // literals
                        case 'e':
                        case 'E':
                        case '.':
                            break;
                        default:
                            // read in the character so we have something to error on
                            readChar();

                            // throw an error indicating that the base code is unknown
                            throwScanError(std::format("unknown interger base code: {}", ahead.value()));
                            break;
                    }
                }
            }
        }

        // these flags determine what kind of number literal is generated and
        // are used to maintain the state regardingly that literal as it is
        // being read in by the scanner.
        bool isFloating = false, encounteredExp = false, expectingNeg = false;
        bool isUnsigned = false, isLong = false;

        while (auto ahead = peekChar()) {
            // suffixes can only occur once per and at the end
            if (isUnsigned && isLong)
                break;
            else if (isUnsigned) {
                if (ahead == 'l') {
                    readChar();
                    isLong = true;
                    continue;
                }
                    
                break;
            } else if (isLong) {
                if (ahead == 'u') {
                    readChar();
                    isUnsigned = true;
                    continue;
                }
                
                break;
            }

            switch (base) {
                // binary literals
                case 2:
                    if (ahead == '0' || ahead == '1')
                        readChar();
                    else
                        goto loopexit;
                    break;
                // octal literals
                case 8:
                    if ('0' <= ahead && ahead < '8')
                        readChar();
                    else
                        goto loopexit;
                    break;
                // hex literals
                case 16:
                    if (isdigit(ahead.value()) || 'a' <= ahead && ahead <= 'f' || 'A' <= ahead && ahead <= 'F')
                        readChar();
                    else
                        goto loopexit;
                    break;
                // base 10 literals are the only literals that can also be
                // floating point numbers
                case 10:
                    if (isdigit(ahead.value()))
                        readChar();
                    else {
                        switch (ahead.value()) {
                            case 'e':
                            case 'E':
                                readChar();

                                if (encounteredExp) 
                                    throwScanError("floating literal cannot contain multiple exponents");

                                encounteredExp = true;
                                isFloating = true;
                                expectingNeg = true;
                                continue;
                            case '.':
                                readChar();

                                // if it is floating then either the decimal is misplaced after the exponent
                                // or multiple decimals occur in the literal: this is invalid in either case
                                if (isFloating)
                                    throwScanError("unexpected decimal point in float literal");

                                isFloating = true;
                                break;
                            case 'l':
                                readChar();

                                if (isFloating)
                                    throwScanError("float literal cannot be marked long");

                                isLong = true;
                                break;
                            case 'u':
                                readChar();

                                if (isFloating)
                                    throwScanError("float literal cannot be marked unsigned");

                                isUnsigned = true;   
                                break;
                            case '-':
                                // an unexpected negative is not necessarily an
                                // exit condition: eg. `5-2` is a valid Chai
                                // expression; erroring here would cause such an
                                // expression to be rejected
                                if (expectingNeg) {
                                    readChar();
                                    expectingNeg = false;
                                    continue;
                                }

                                // because using the same keyword to exit switch
                                // cases and loops was an *AMAZING* structured
                                // programming innovation /s
                                goto loopexit;  
                            default:
                                goto loopexit;                             
                        }
                    }

                    // clear expecting negative if we reach here
                    expectingNeg = false;
            }
        }

    loopexit:
        if (isFloating)
            return makeToken(TokenKind::FloatLiteral);
        else if (isUnsigned || isLong || base != 10)
            return makeToken(TokenKind::IntLiteral);
        else
            return makeToken(TokenKind::NumLiteral);
    }

    // scanStdStringLit scans in a standard (double quoted) string literal
    // assuming the leading double quote has already been read in.
    Token Scanner::scanStdStringLit() {
        while (auto ahead = readChar()) {
            switch (ahead.value()) {
                case '\n':
                    throwScanError("standard string cannot contain newlines");
                    break;
                case '\"':
                    return makeToken(TokenKind::StringLiteral);
                case '\\':
                    readEscapeSequence();
                    break;
            }
        }

        throwScanError("expected closing double quote for standard string literal");
    }

    // scanRawStringLit scans in a raw/multiline (backtick quoted) string
    // literal assuming the leading backtick has already been read in.
    Token Scanner::scanRawStringLit() {
        while (auto ahead = readChar()) {
            if (ahead == '`')
                return makeToken(TokenKind::StringLiteral);
        }

        throwScanError("expected closing backtick for raw string literal");
    }

    // -------------------------------------------------------------------------- //

    // makeToken creates a new token based on the contents of the token buffer
    Token Scanner::makeToken(TokenKind kind) {
        Token token = {.kind = kind, .value=tokBuff, .line=line, .col=col};
        tokBuff.clear();

        return token;
    }

    // throwScanError throws a new error regarding the current token being scanned
    void Scanner::throwScanError(const std::string& msg) {
        throw CompileMessage(msg, {.startLine=tokStartLine, .startCol=tokStartCol, .endLine=line, .endCol=col}, fpath);
    }

    // -------------------------------------------------------------------------- //

    // readChar attempts to get a new character from the input stream.  It also
    // updates the position of the scanner and adds the character to the current
    // token buffer.
    std::optional<char> Scanner::readChar() {
        char c = 0;
        if (!file.get(c))
            return {};

        updatePosition(c);
        tokBuff.push_back(c);
        return c;
    }

    // skipChar attempts to get a new character from the input stream.  If it
    // succeeds, the scanner's position is updated and the character is
    // discarded.
    bool Scanner::skipChar() {
        char c = 0;
        if (!file.get(c))
            return false;

        updatePosition(c);
        return true;
    }

    // peekChar attempts to look one char ahead in the input stream.  If it
    // succeeds the char is returned.  In any case, this function has no effect
    // on the scanner's state -- it acts as a lookahead.
    std::optional<char> Scanner::peekChar() {
        int c = file.peek();
        if (c == EOF)
            return {};

        return (char)c;
    }

    // updatePosition updates the position of the scanner based on a char that
    // was just read in.
    void Scanner::updatePosition(char c) {
        // whenever the token buffer is of length zero and a character is
        // written to it, the start position is reset
        if (tokBuff.length() == 0) {
            tokStartLine = line;
            tokStartCol = col;
        }

        switch (c) {
            case '\n':
                line++;
                col = 1;
                break;
            case '\t':
                col += 4;
                break;
            default:
                col++;
                break;
        }
    }
}